fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/mixin/event/interaction/MixinBlockItem.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.event.interaction;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.item.BlockItem;
+import net.minecraft.item.ItemPlacementContext;
+
+import net.fabricmc.fabric.api.event.player.PlayerBlockPlaceEvents;
+
+@Mixin(BlockItem.class)
+public class MixinBlockItem {
+	@Inject(method = "place(Lnet/minecraft/item/ItemPlacementContext;Lnet/minecraft/block/BlockState;)Z", at = @At("HEAD"), locals = LocalCapture.CAPTURE_FAILHARD, cancellable = true)
```

Comment:
I do not think this is the right place to fire the event. As the method is on block, someone may override this method and ignore the event completely.

The problem is there is really no other way to distinguish between a block placement and opening a furnace if we were to inject where this was called.

We might be able to get away with this inside of `ServerPlayerInteractionManager#interactBlock`:

```java
// Inject before ItemStack#useOnBlock, set `require` to 2 so we can fail if this changes in a future version for validating existing code
if (stack.getItem() instanceof BlockItem) {
    // TODO: Fire event
    if (!result) {
        // Cancel and fail
    }
}
```

and then to check the placement was successful for the after event, inject before/after `Criteria.ITEM_USED_ON_BLOCK.test(player, blockPos, itemStack);` and fire the `after` event.

Also this is also pretty much the correct place for any future events related to item interaction, just not sure how it should be handled relatively to block place yet, but that is a discussion for that other pull request.

Reply:
Totally agree, however have noticed that the UseBlockCallback event is here and seems to be a basic version of what we actually want, maybe we deprecate that callback and use the newer one with the before and after methods?

Reply:
The existing callbacks need to be thought about since the codepath is a bit weird, such as returning`SUCCESS` in one of those events basically says, ignore what every other mod says, just do the codepath. Refactors and deprecation of those events will probably be in a v1 migration of this module in the future.

Reply:
I have rewritten the new event to point to the new mixin location